freeipa/src/main/java/com/sequenceiq/freeipa/service/stack/RebootInstancesService.java

```diff
@@ -0,0 +1,111 @@
+
+    @Inject
+    private StackService stackService;
+
+    @Inject
+    private FreeIpaFlowManager flowManager;
+
+    @Inject
+    private FreeIpaHealthDetailsService healthDetailsService;
+
+    @Inject
+    private InstanceMetaDataService instanceMetaDataService;
+
+    private Map<String, InstanceStatus> getInstanceHealth(String accountId, RebootInstancesRequest request) {
+        if (!request.isForceReboot()) {
+            return healthDetailsService.getHealthDetails(request.getEnvironmentCrn(), accountId).getNodeHealthDetails().stream()
+                    .collect(Collectors.toMap(NodeHealthDetails::getInstanceId, NodeHealthDetails::getStatus));
+        }
+        return Collections.emptyMap();
+    }
+
+    private Set<String> getInvalidInstances(Collection<String> allInstances, RebootInstancesRequest request) {
+        Consumer<String> logit = id -> {
+            if (!allInstances.contains(id)) {
+                LOGGER.info("Unknown instance ID {} in request", id);
+            }
+        };
+        return Stream.ofNullable(request.getInstanceIds()).flatMap(Collection::stream)
+                .peek(logit)
+                .filter(id -> !allInstances.contains(id))
+                .collect(Collectors.toSet());
+    }
+
+    private Map<String, InstanceMetaData> getInstancesToReboot(Map<String, InstanceMetaData> allInstances,
+            List<String> instanceIds, Map<String, InstanceStatus> health) {
+        Consumer<Map.Entry> logit = entry -> {
+            if (health.get(entry.getKey()) != null && health.get(entry.getKey()).isAvailable()) {
+                LOGGER.info("Not rebooting AVAILABLE instance {} because force reboot was not selected.", entry.getKey());
+            }
+        };
+        return allInstances.entrySet().stream()
+                .peek(logit)
+                .filter(entry -> instanceIds == null || instanceIds.isEmpty() || instanceIds.contains(entry.getKey()))
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+    }
+
+    private Map<String, InstanceMetaData> getAllInstancesMap(Stack stack) {
+        return stack.getInstanceGroups().stream().flatMap(instanceGroup -> instanceGroup.getInstanceMetaData().stream())
+                .collect(Collectors.toMap(InstanceMetaData::getInstanceId, Function.identity()));
+    }
+
+    public void rebootInstances(String accountId, RebootInstancesRequest request) {
+        // If no instance passed in request, reboot all bad instances
+        // If instances passed in request, reboot all valid passed bad instances
+        // If force and instances passed in request, reboot all valid passed instances
+        // If force and no instances passed in request, reboot all instances
+        Stack stack = stackService.getByEnvironmentCrnAndAccountIdWithLists(request.getEnvironmentCrn(), accountId);
+        MDCBuilder.buildMdcContext(stack);
+        Map<String, InstanceMetaData> allInstances = getAllInstancesMap(stack);
```

Comment:
could you rename the variable? like `allInstanceByInstanceId`, because I had to lookup the code to know what is in this map

Reply:
fixed